game/groundforces/template_loader.py

```diff
@@ -0,0 +1,361 @@
+from __future__ import annotations
+
+import itertools
+import logging
+import pickle
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Iterator, Optional, Any
+
+import dcs
+import yaml
+from dcs import Point
+from dcs.unitgroup import StaticGroup
+
+from game import persistency
+from game.version import VERSION
+from game.data.groups import GroupRole, GroupTask
+from game.data.units import UnitClass
+from game.groundforces.template import (
+    GroundObjectTemplate,
+    GroupTemplate,
+    UnitTemplate,
+    AntiAirTemplate,
+    BuildingTemplate,
+    NavalTemplate,
+    GroundForceTemplate,
+    DefensesTemplate,
+)
+from game.profiling import logged_duration
+
+TEMPLATE_DIR = "resources/templates/"
+TEMPLATE_DUMP = "Liberation/templates.p"
+
+
+@dataclass
+class GroupTemplateMapping:
+    # The group name used in the template.miz
+    name: str
+
+    # Defines if the group is required for the template or can be skipped
+    optional: bool = False
+
+    # All static units for the group
+    statics: list[str] = field(default_factory=list)
+
+    # Defines to which tgo group the groupTemplate will be added
+    # This allows to merge groups back together. Default: Merge all to group 1
+    group: int = field(default=1)
+
+    # Randomization settings. If left empty the template will be generated with the
+    # exact values (amount of units and unit_type) defined in the template.miz
+    # How many units the template should generate. Will be used for randomization
+    unit_count: list[int] = field(default_factory=list)
+    # All unit types the template supports.
+    unit_types: list[str] = field(default_factory=list)
+    # All unit classes the template supports.
+    unit_classes: list[UnitClass] = field(default_factory=list)
+    alternative_classes: list[UnitClass] = field(default_factory=list)
```

Comment:
Alternatives are equally weighted fallbacks if none of the entries in `unit_classes` are valid for the faction?

Is `unit_classes` the right place for the randomization to happen, or should that just be a priority ordered list and `alternative_classes` can go away? Perhaps I'm thinking of this backward, but we're matching templates to groups, not filling templates _with_ groups, right? In other words, the campaign designer (or `start_generator`) picks a group first, and picks a template based on that group, rather than picking a _template_ that gets auto-filled by some matching group. If that's the case I don't understand what purpose the randomness serves (this isn't my usual knee-jerk reaction to randomness, I genuinely don't understand how it has any observable effect).

Reply:
The alternative_classes will only be used when the faction has none of the unit_classes defined accessible. It is a fallback which currently is only used for the EWR Layout to allow SR and STR to be fallback Radars to provide the EWR Role. I agree that this can be implemented better.
I'll write it on my list of improvements anyway. Its actually not so trivial as for example adding a fixed priority order into the unit_classes list would change the overall concept massively as the current approach is that they are equally weighted and randomization takes over.
With more use cases i would implement a better approach. Currently as is said its only purpose is the EWR template.

A 2nd Template with SR and STR as class will not help as it would be similar ranked than the one with explicit EWR capable radars. Also adding the two classes there will mess up the EWRs.